controller/scheduler-v2/scheduler.go

```diff
@@ -339,7 +337,13 @@ func (s *Scheduler) HandleFormationChange(ef *ct.ExpandedFormation) {
 		s.sendEvent(NewEvent(EventTypeFormationChange, err, nil))
 	}()
 
-	log := logger.New("fn", "HandleFormationChange", "app.id", ef.App.ID, "release.id", ef.Release.ID)
+	log := logger.New("fn", "HandleFormationChange")
+	if ef.App != nil {
+		log = log.New("app.id", ef.App.ID)
+	}
+	if ef.Release != nil {
+		log = log.New("release.id", ef.Release.ID)
+	}
```

Comment:
When `StreamFormations` is called, it first sends all the formations updated since the `since` argument, so if you get disconnected, you can reconnect and get the missed changes. To tell when you are up-to-date, a sentinel value is sent (i.e. [here](https://github.com/flynn/flynn/blob/master/controller/formation.go#L285))